ethexe/contracts/src/MirrorAbi.sol

```diff
@@ -0,0 +1,177 @@
+    modifier retrievingValue(uint128 value) {
+        if (value != 0) {
+            address routerAddr = router;
+            bool success = _wvara(routerAddr).transferFrom(msg.sender, routerAddr, value);
+            require(success, "failed to transfer non-zero amount of WVara from source to router");
+        }
+        _;
+    }
+
+    // TODO (breathx): terminated programs compute threshold must always be treated as balance-enough.
+    /// @dev Functions marked with this modifier can be called only after the program is terminated.
+    modifier whenTerminated() {
+        require(inheritor != address(0), "program is not terminated");
+        _;
+    }
+
+    /// @dev Functions marked with this modifier can be called only after the initializer has created the init message.
+    modifier whenInitMessageCreated() {
+        require(nonce > 0, "initializer hasn't created init message yet");
+        _;
+    }
+
+    /// @dev Functions marked with this modifier can be called only after the initializer has created the init message or from the initializer (first access).
+    modifier whenInitMessageCreatedOrFromInitializer() {
+        require(
+            nonce > 0 || msg.sender == initializer,
+            "initializer hasn't created init message yet; and source is not initializer"
+        );
+        _;
+    }
+
+    /// @dev Functions marked with this modifier can be called only if the program is active.
+    modifier whileActive() {
+        require(inheritor == address(0), "program is terminated");
+        _;
+    }
+
+    function sendMessage(bytes calldata /*_payload*/, uint128 _value)
+        external
+        whileActive
+        whenInitMessageCreatedOrFromInitializer
+        retrievingValue(_value)
+        returns (bytes32 resultData)
+    {
+        bytes memory result = _mirrorImplDelegatecall();
+
+        assembly {
+            resultData := mload(result)
+        }
+    }
+
+    function sendReply(bytes32 /*_repliedTo*/, bytes calldata /*_payload*/, uint128 _value)
+        external
+        whileActive
+        whenInitMessageCreated
+        retrievingValue(_value)
+    {
+        _mirrorImplDelegatecall();
+    }
+
+    function claimValue(bytes32 /*_claimedId*/) external whenInitMessageCreated {
+        _mirrorImplDelegatecall();
+    }
+
+    function executableBalanceTopUp(uint128 _value) external whileActive retrievingValue(_value) {
+        _mirrorImplDelegatecall();
+    }
+
+    function transferLockedValueToInheritor() public whenTerminated {
+        _mirrorImplDelegatecall();
+    }
+
+    // NOTE (breathx): value to receive should be already handled in router.
+    function performStateTransition(Gear.StateTransition calldata /*_transition*/) external onlyRouter returns (bytes32 resultData) {
+        bytes memory result = _mirrorImplDelegatecall();
+
+        assembly {
+            resultData := mload(result)
+        }
+    }
+
+    receive() external payable {
+        revert("");
```

Comment:
`receive` does `revert` by default

Reply:
there is a warning w/o `receive` function
`This contract has a payable fallback function, but no receive ether function. Consider adding a receive ether function.`

Reply:
ok, change it to `revert()`